src/Network/Wai/Middleware/OpenApi/PathMap.hs

```diff
@@ -0,0 +1,59 @@
+module Network.Wai.Middleware.OpenApi.PathMap
+  ( PathMap
+  , fromOpenApi
+  , lookup
+  ) where
+
+import Prelude hiding (lookup)
+
+import Control.Lens ((^?))
+import Data.Bifunctor (first)
+import Data.ByteString (ByteString)
+import Data.ByteString.Char8 qualified as BS8
+import Data.HashMap.Strict.InsOrd qualified as IHashMap
+import Data.Map.Strict (Map)
+import Data.Map.Strict qualified as Map
+import Data.OpenApi (OpenApi, PathItem)
+import Data.OpenApi qualified as OpenApi
+import System.FilePath.Posix qualified as Posix
+
+newtype PathMap = PathMap
+  { unwrap :: Map TemplatedPath PathItem
+  }
+
+fromOpenApi :: OpenApi -> PathMap
+fromOpenApi spec =
+  PathMap $ case spec ^? OpenApi.paths of
+    Nothing -> Map.empty
+    Just ps -> Map.fromList $ map (first toTemplatedPath) $ IHashMap.toList ps
+
+lookup :: ByteString -> PathMap -> Maybe PathItem
+lookup p pm = Map.lookup (toTemplatedPath $ BS8.unpack p) pm.unwrap
+
+newtype TemplatedPath = TemplatedPath
+  { _unwrap :: [TemplatedPathComponent]
+  }
+  deriving stock (Eq, Ord)
+
+toTemplatedPath :: FilePath -> TemplatedPath
+toTemplatedPath =
+  TemplatedPath
+    . map toTemplatedPathComponent
+    . Posix.splitDirectories
+
+data TemplatedPathComponent
+  = Exact FilePath
+  | ParameterValue
+
+instance Eq TemplatedPathComponent where
+  Exact l == Exact r = l == r
+  _ == _ = True
```

Comment:
[Eq has a transitivity law](https://hackage.haskell.org/package/base-4.20.0.1/docs/Data-Eq.html#t:Eq) that's violated where

- `x = Exact something`
- `y = ParameterValue`
- `z = Exact somethingElse`

Reply:
Yeah, fair enough.